### PR TITLE
Link TextField component implementation

### PR DESCRIPTION
**Pull request by Neckow**
#43

### CODE REVIEW

File: RunningOrder/Shared/Components/InlineEditableLink.swift [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                if isEditing {
+                    VStack {
+                        BorderedTextField(placeholder: "Label", value: $value.label)
+                        BorderedTextField(placeholder: "Url", value: $value.url)
+                    }
+                    Spacer()
+                    Button(action: {
+                        if !isFieldsEmpty {
+                            self.isEditing = false
+                            value.url = formatURL(content: value.url)
```
Comment (zarghol) [inline RunningOrder/Shared/Components/InlineEditableLink.swift]: create a new logic

File: RunningOrder/Shared/Components/InlineEditableLinkList+Logic.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import Foundation
+import Combine
+import SwiftUI
+
+extension InlineEditableLinkList {
+    final class Logic: ObservableObject {
+        @Binding var values: [LinkEntity]
+
+        internal init(values: Binding<[LinkEntity]>) {
```
Comment (zarghol) [inline RunningOrder/Shared/Components/InlineEditableLinkList+Logic.swift]: remove internal

File: RunningOrder/Shared/Components/StyledFocusableTextField.swift [hunk 3]
```diff
@@ -24,10 +24,11 @@ struct StyledFocusableTextField: View {
     ///   - placeholder: the textfield placeholder
     ///   - value: the binding to the textfield value
     ///   - onCommit: the action to perform when the user hits the return key or when the component looses the focus
-    init(_ placeholder: String, value: Binding<String>, onCommit: @escaping () -> Void) {
+    init(_ placeholder: String, value: Binding<String>, initialFocus: Bool = false, onCommit: @escaping () -> Void) {
```
Comment (zarghol) [inline RunningOrder/Shared/Components/StyledFocusableTextField.swift]: is this change still used, as we created a new component

File: RunningOrder/Story/Models/Link.swift [hunk 4]
```diff
@@ -8,15 +8,23 @@
 
 import Foundation
 
-struct Link {
+struct LinkEntity: Identifiable {
```
Comment (zarghol) [inline RunningOrder/Story/Models/Link.swift]: maybe rename it Link again ?

File: RunningOrder/Story/Models/Configuration.swift [hunk 5]
```diff
@@ -14,6 +14,7 @@ struct Configuration {
     var features: [String] = []
     var indicators: [String] = []
     var identifiers: [String] = []
+    var links: [LinkEntity] = []
```
Comment (zarghol) [inline RunningOrder/Story/Models/Configuration.swift]: remove from here

File: RunningOrder/Story/Models/StoryInformation.swift [hunk 6]
```diff
@@ -13,7 +13,6 @@ struct StoryInformation {
     let storyId: Story.ID
 
     var configuration = Configuration()
-    var links: [Link] = []
```
Comment (zarghol) [inline RunningOrder/Story/Models/StoryInformation.swift]: revert please

**Review by zarghol — CHANGES_REQUESTED** [review]

File: RunningOrder/Shared/Components/InlineEditableLink.swift [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    @State private var isDoneButtonEnabled = true
+
+    /// - Parameters:
+    ///   - value: A binding to a Link Entity which referers to the Link url and label
+    init(value: Binding<LinkEntity>) {
+        self._value = value
+        self._isEditing = State(initialValue: value.wrappedValue.label.isEmpty)
+    }
+
+    var isFieldsEmpty: Bool {
```
Comment (ghiitais) [inline RunningOrder/Shared/Components/InlineEditableLink.swift]: isFieldEmpty

File: RunningOrder/Story/Models/Configuration.swift (hunk 5 above)
Comment (ghiitais) [inline RunningOrder/Story/Models/Configuration.swift]: Why not just Link ? Instead of LinkEntity?

Comment (Neckow) [reply RunningOrder/Story/Models/Configuration.swift]: Cause there was a conflict with Swift UI class "Link" but I fixed it (changes in the upcoming commit)

**Review by ghiitais — APPROVED** [review]

File: RunningOrder/Story/Models/Configuration.swift (hunk 5 above)
Comment (ghiitais) [reply RunningOrder/Story/Models/Configuration.swift]: aaaah

File: RunningOrder/Shared/Components/InlineEditableLink.swift [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+                            logic.value.url = logic.formatURL(content: logic.value.url)
+                        }
+                    }, label: {
+                        Text("Done")
+                    })
+                    .buttonStyle(PlainButtonStyle())
+                    .foregroundColor(.accentColor)
+                    .disabled(logic.isFieldEmpty)
+                } else {
+                    HStack {
```
Comment (zarghol) [inline RunningOrder/Shared/Components/InlineEditableLink.swift:45]: I don't understand really this part : 
in the case of not editing, you have a HStackView, that can be empty if there is no formattedURL ?

File: RunningOrder/Shared/Components/InlineEditableLink.swift [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+                if isEditing {
+                    VStack {
+                        BorderedTextField(placeholder: "Label", value: logic.$value.label)
+                        BorderedTextField(placeholder: "Url", value: logic.$value.url)
+                    }
+                    Spacer()
+                    Button(action: {
+                        if !logic.isFieldEmpty {
+                            self.isEditing = false
+                            logic.value.url = logic.formatURL(content: logic.value.url)
```
Comment (zarghol) [inline RunningOrder/Shared/Components/InlineEditableLink.swift:36]: all of this should be in the logic method

File: RunningOrder/Shared/Components/InlineEditableLink.swift [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    ///   - value: A binding to a Link Entity which referers to the Link url and label
+    init(value: Binding<Link>) {
+        self._isEditing = State(initialValue: value.wrappedValue.label.isEmpty)
+        self.logic = Logic(value: value)
+    }
+
+    var body: some View {
+        VStack(alignment: .leading) {
+            HStack {
+                if isEditing {
```
Comment (zarghol) [inline RunningOrder/Shared/Components/InlineEditableLink.swift:27]: I think it could be re-worked :
```swift
if let url = logic.value.formattedURL, ! isEditing {
   SwiftUI.Link(logic.value.label, destination: url)
   Spacer()
   if hovered {
      Button("Edit")  { self.isEditing = true }
          .buttonStyle(PlainButtonStyle())
          .foregroundColor(.accentColor)
   }
} else {
   VStack {
       BorderedTextField(placeholder: "Label", value: logic.$value.label)
       BorderedTextField(placeholder: "Url", value: logic.$value.url)
   }
   Spacer()
   Button("Done", action: logic.doneAction)
      .buttonStyle(PlainButtonStyle())
      .foregroundColor(.accentColor)
      .disabled(logic.isFieldEmpty)
}
```

Comment (zarghol) [reply RunningOrder/Shared/Components/InlineEditableLink.swift:27]: not now maybe, but later

File: RunningOrder/Shared/Components/InlineEditableLinkList.swift [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+                )
+                .foregroundColor(.blue)
+                .buttonStyle(InlineButtonStyle())
+            }
+            .padding(.bottom, 10)
+
+            ForEach(logic.values.indices, id: \.self) { index in
+                HStack {
+                    InlineEditableLink(
+                        value: Binding(
```
Comment (zarghol) [inline RunningOrder/Shared/Components/InlineEditableLinkList.swift:43]: this binding creation could be part of the logic maybe 🤔

File: RunningOrder/Shared/Components/StyledFocusableTextField.swift [hunk 12]
```diff
@@ -13,7 +13,7 @@ struct StyledFocusableTextField: View {
 
     let placeholder: String
 
-    @State private var isFocused = false
+    @State private var isFocused: Bool = false
```
Comment (zarghol) [inline RunningOrder/Shared/Components/StyledFocusableTextField.swift:16]: why ?

File: RunningOrder/Story/Models/Link.swift [hunk 13]
```diff
@@ -8,15 +8,23 @@
 
 import Foundation
 
-struct Link {
+struct Link: Identifiable {
+    var id = UUID()
```
Comment (zarghol) [inline RunningOrder/Story/Models/Link.swift:12]: As it was before LinkEntity, I didn't see these changes :
why did you add uuid ? please remove it, we don't need it.

please add the conformance to identifiable in an extension, as other conformance. the id needed for identifiable could be something simpler such as the label. (does it have sense to have multiple links with the same label ?)

File: RunningOrder/Story/Models/StoryInformation.swift [hunk 14]
```diff
@@ -35,10 +35,9 @@ struct StoryInformation {
 
     private var videoExtension: String?
 
-    init(storyId: Story.ID, configuration: Configuration = Configuration(), links: [Link] = [], steps: [String] = [], videoUrl: URL? = nil) {
+    init(storyId: Story.ID, configuration: Configuration = Configuration(), steps: [String] = [], videoUrl: URL? = nil) {
```
Comment (zarghol) [inline RunningOrder/Story/Models/StoryInformation.swift:38]: could you please revert this to enable the possibility to add links at init ?

File: RunningOrder/Story/Views/ConfigurationView.swift [hunk 15]
```diff
@@ -28,18 +28,7 @@ struct ConfigurationView: View {
...
-                    .padding(.leading, 10)
-
-                InlineEditableList(
-                    title: "Links",
-                    values: Binding<[String]>(
-                        get: { self.storyInformation.links.map { $0.label } },
-                        set: { values in self.storyInformation.links = values.map { Link(value: $0) } }
-                    )
-                )
+                InlineEditableLinkList(title: "Link", values: $storyInformation.links)
```
Comment (zarghol) [inline RunningOrder/Story/Views/ConfigurationView.swift:31]: Links ?

**Review by zarghol — APPROVED** [review]
There is still changes to do, but maybe we can do it independently of this MR

File: RunningOrder/Story/Models/StoryInformation.swift (hunk 14 above)
Comment (Neckow) [reply RunningOrder/Story/Models/StoryInformation.swift:38]: Already resolved
